### PR TITLE
 Implementation of Average Marginal Ranking Loss Function

### PR DESCRIPTION
**Pull request by jkauerl**
# Pull Request Template

## Description

PR including the implementation for the Marginal Ranking Loss Function of this issue #559 . It creates a new file called `marginal_ranking.rs` with the function called `mrg_ranking_loss`. This implementation is inspired on the documentation of PyTorch of [Marginal Ranking](https://pytorch.org/docs/stable/generated/torch.nn.MarginRankingLoss.html).

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

PD: Im new to Rust and Open Source so I would love a clarification regarding the items I did not checked. Also let me know if I should squash the commits. Last PR had a wrong branch name #733.

### CODE REVIEW

**Comment by jkauerl** [comment]
Made some pretty major updated following guidlines I have recieved before.

File: src/machine_learning/loss_function/marginal_ranking.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+/// It returns the average loss by dividing the `total_loss` by total no. of
+/// elements.
+///
+/// Pytorch implementation:
+/// https://pytorch.org/docs/stable/generated/torch.nn.MarginRankingLoss.html
+/// https://gombru.github.io/2019/04/03/ranking_loss/
+/// https://vinija.ai/concepts/loss/#pairwise-ranking-loss
+///
+
+pub fn mrg_ranking_loss(
```
Comment (vil02) [inline src/machine_learning/loss_function/marginal_ranking.rs:22]:
```suggestion
pub fn marginal_ranking_loss(
```

File: src/machine_learning/loss_function/marginal_ranking.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    y_true: f64,
+) -> Result<f64, MarginalRankingLossError> {
+    if x_first.len() != x_second.len() {
+        return Err(MarginalRankingLossError::InputsHaveDifferentLength);
+    }
+    if x_first.is_empty() {
+        return Err(MarginalRankingLossError::EmptyInputs);
+    }
+    if margin < 0.0 {
+        return Err(MarginalRankingLossError::InvalidValues);
```
Comment (vil02) [inline src/machine_learning/loss_function/marginal_ranking.rs:35]: What do you think about introducing a new error and:
```suggestion
        return Err(MarginalRankingLossError::NegativeMargin);
```

File: src/machine_learning/loss_function/marginal_ranking.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+mod tests {
+    use super::*;
+
+    macro_rules! test_with_wrong_inputs {
+        ($($name:ident: $inputs:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (x_first, x_second, margin, y_true, expected) = $inputs;
+                    assert_eq!(mrg_ranking_loss(&x_first, &x_second, margin, y_true), expected);
```
Comment (vil02) [inline src/machine_learning/loss_function/marginal_ranking.rs:66]: The _error handing_ is _symmetric_ in this case, therefore one could use something like:
```suggestion
                    let (vec_a, vec_b, margin, y_true, expected) = $inputs;
                    assert_eq!(mrg_ranking_loss(&vec_a, &vec_b, margin, y_true), expected);
                    assert_eq!(mrg_ranking_loss(&vec_b, &vec_a, margin, y_true), expected);
```

File: src/machine_learning/loss_function/marginal_ranking.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    }
+    if y_true != 1.0 && y_true != -1.0 {
+        return Err(MarginalRankingLossError::InvalidValues);
+    }
+
+    let mut total_loss: f64 = 0.0;
+    for (f, s) in x_first.iter().zip(x_second.iter()) {
+        let loss: f64 = (margin - y_true * (f - s)).max(0.0);
+        total_loss += loss;
+    }
```
Comment (vil02) [inline src/machine_learning/loss_function/marginal_ranking.rs:45]:
```suggestion
    let total_loss: f64 = x_first
        .iter()
        .zip(x_second.iter())
        .map(|(f, s)| (margin - y_true * (f - s)).max(0.0))
        .sum();
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/machine_learning/loss_function/marginal_ranking.rs (hunk 2 above)
Comment (jkauerl) [reply src/machine_learning/loss_function/marginal_ranking.rs:35]: That sounds perfect, will add that.

File: src/machine_learning/loss_function/marginal_ranking.rs (hunk 3 above)
Comment (jkauerl) [reply src/machine_learning/loss_function/marginal_ranking.rs:66]: Makes sense, will add that to the code.

File: src/machine_learning/loss_function/marginal_ranking.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    if x_first.is_empty() {
+        return Err(MarginalRankingLossError::EmptyInputs);
+    }
+    if margin < 0.0 {
+        return Err(MarginalRankingLossError::NegativeMargin);
+    }
+    if y_true != 1.0 && y_true != -1.0 {
+        return Err(MarginalRankingLossError::InvalidValues);
+    }
```
Comment (vil02) [inline src/machine_learning/loss_function/marginal_ranking.rs:39]:
```suggestion
    check_input(x_first, x_second, margin, y_true)?;
```
where
```rust
fn check_input(
    x_first: &[f64],
    x_second: &[f64],
    margin: f64,
    y_true: f64,
) -> Result<(), MarginalRankingLossError> {
    if x_first.len() != x_second.len() {
        return Err(MarginalRankingLossError::InputsHaveDifferentLength);
    }
    if x_first.is_empty() {
        return Err(MarginalRankingLossError::EmptyInputs);
    }
    if margin < 0.0 {
        return Err(MarginalRankingLossError::NegativeMargin);
    }
    if y_true != 1.0 && y_true != -1.0 {
        return Err(MarginalRankingLossError::InvalidValues);
    }

    Ok(())
}
```

File: src/machine_learning/loss_function/marginal_ranking.rs
Comment (vil02) [inline src/machine_learning/loss_function/marginal_ranking.rs:1]: I am slightly confused by the naming: why _marginal_? The links, which you have provided are about _margin ranking loss_. Furthermore you compute the average here. What do you think about using the name `average_margin_ranking_loss` or something similar.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/machine_learning/loss_function/marginal_ranking.rs
Comment (jkauerl) [reply src/machine_learning/loss_function/marginal_ranking.rs:1]: From what I understand is the same formular but written in a slightly different manner. I am open to change the name of the function for it to be more precise, but every loss function calculates the average and so I though that was the standard?

File: DIRECTORY.md [hunk 6]
```diff
@@ -155,6 +155,7 @@
       * [Hinge Loss](https://github.com/TheAlgorithms/Rust/blob/master/src/machine_learning/loss_function/hinge_loss.rs)
       * [Huber Loss](https://github.com/TheAlgorithms/Rust/blob/master/src/machine_learning/loss_function/huber_loss.rs)
       * [Kl Divergence Loss](https://github.com/TheAlgorithms/Rust/blob/master/src/machine_learning/loss_function/kl_divergence_loss.rs)
+      * [Marginal Ranking Loss](https://github.com/TheAlgorithms/Rust/blob/master/src/machine_learning/loss_function/marginal_ranking.rs)
```
Comment (vil02) [inline DIRECTORY.md:158]:
```suggestion
      * [Marginal Ranking Loss](https://github.com/TheAlgorithms/Rust/blob/master/src/machine_learning/loss_function/average_margin_ranking_loss.rs)
```

**Review by vil02 — APPROVED** [review]
